nixos/modules/config/ldap.nix

```diff
@@ -214,6 +214,9 @@ in
       if cfg.daemon.enable then nss_pam_ldapd else nss_ldap
     );
 
+    # Let pwdutils tools query nslcd through lbnss_ldap.so.
+    environment.sessionVariables.LD_LIBRARY_PATH = [ "${pkgs.nss_pam_ldapd}/lib" ];
```

Comment:
For nss modules we usually add those nscd's library path instead.

Reply:
oh right, thank you Mic92! I've amended the pull request (as suggested by https://nixos.org/nixpkgs/manual/#idm140737315696608 )
The result seems ok.
```
# echo $LD_LIBRARY_PATH
/nix/store/rx76y57vl69ynnig325pid8jb0466298-systemd-239/lib:/nix/store/lm4vib2nl74z5hpsn11wfw2jxvmsgrwb-nss-pam-ldapd-0.9.7/lib
```
I still don't know though how to let `passwd` actually change a password through LDAP.

Reply:
@flokli
Thanks. So, I am now able to change the LDAP password with `passwd`.
In `nixpkgs/nixos/modules/security/pam.nix` I had to:
```
           # Password management.
-          password requisite pam_unix.so nullok sha512
+          password sufficient pam_unix.so nullok sha512
```
I don't know if this fix is "secure", nor how to apply it simply without modifying Nixpkgs directly. In the same file, one can read:
```
# !!! TODO: move the LDAP stuff to the LDAP module, and the
# Samba stuff to the Samba module.  This requires that the PAM
# module provides the right hooks.
```

FYI, in the LDAP database, not only had I to give `write` and `auth` access to `userPassword`, but also `read` access to `self` on the `ou=posix,…` subtree, because when a user `$USER` runs `passwd $USER`, `nslcd` binds to the LDAP server as `$USER` with the current password and then do some searches on that subtree:
```
olcAccess: to attrs=userPassword
  by self write
  by anonymous auth
  by dn="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" write
  by * none
olcAccess: to attrs=shadowLastChange
  by self write
  by dn="gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth" write
  by * none
olcAccess: to dn.sub="ou=posix,dc=example,dc=coop"
  by self read
  by dn="gidNumber=${toString groups.nslcd.gid}+uidNumber=${toString users.nslcd.uid},cn=peercred,cn=external,cn=auth" read
```

Concerning the NixOS configuration of `nslcd`, here is my current one:
```
config = { 
  users.ldap = { 
    enable = true;
    server = "ldapi:///";
    base = "ou=posix,dc=example,dc=coop";
    daemon = { 
      enable = true;
      extraConfig = ''
        sasl_mech EXTERNAL
          # NOTE: nslcd cannot use SASL to bind to rootpwmoddn
          # which is the DN used by nslcd when passwd is run by root
          # to change the userPassword of an LDAP user.
          # SEE: https://www.reddit.com/r/linuxadmin/comments/53sxpl/how_do_i_configure_nslcd_to_use_a_sasl_external/d7w9awd/
          # Thus, use: ldappasswd -H ldapi:// -Y EXTERNAL uid=$user,ou=accounts,ou=posix,dc=example,dc=coop
      '';
    };
  };
};
```
Note also that the Reddit comment mentioned above recommends `sssd` as a better alternative to `nslcd`.